### ISSUE TITLE
Tests pass when run from AS 2.3.3, but hang for a while and then fail when run from Gradle

### ISSUE THREAD

**Issue by autonomousapps** [kotlintest/kotlintest#226]
As the title says. I have over 1000 unit tests in the project, and when I run them all from the command line with the command `./gradlew testRegularDebugUnitTest` (regular "flavor"), it hangs at test 896 for a very long time, then fails on one of the tests in spec below. Spec passes in <<100ms when run from AS.

If it matters, we use a variety of testing frameworks in the project. Vanilla JUnit 4, Groovy Spock tests, Robolectric unit tests with ElectricSpecification, and this is out first KotlinTest spec.

AS 2.3.3
Gradle 3.3
Android Build Tools 2.3.3
KotlinTest 2.0.7
Kotlin stdlib 1.1.51

I have a test case that basically looks like this:
```
class LeftNavigationMenuBuilderTest : StringSpec() {
    override val oneInstancePerTest = true // makes no difference

    init {
        "guest users" {
            val menuBuilder = LeftNavigationMenuBuilder.forGuestUser()
            val menu = menuBuilder.buildMenu()
            menu should containsInOrder(GUEST_USER)
        }

        // several others with identical format and different content

    companion object {
        // ITEM{N} are all enum values
        val GUEST_USER = listOf(ITEM1, ITEM2, ITEM3, ITEM4, ITEM5, ITEM6, ITEM7, ITEM8, ITEM9, ITEM10, ITEM11, ITEM12, ITEM13)
        // several others with identical format and different content
    }
}
```

Bizarrely, when I check out the test report, this is what it says:
```
java.lang.AssertionError: expected values must be sorted but was [/* elided, but identical to expected value */]
	at io.kotlintest.matchers.CollectionMatchersKt$containsInOrder$1.test(CollectionMatchers.kt:28)
	at io.kotlintest.matchers.CollectionMatchersKt$containsInOrder$1.test(CollectionMatchers.kt:25)
	at io.kotlintest.matchers.MatchersKt.should(Matchers.kt:71)
	at com.myproject.LeftNavigationMenuBuilderTest$4.invoke(LeftNavigationMenuBuilderTest.kt:33)
	at com.myproject.LeftNavigationMenuBuilderTest$4.invoke(LeftNavigationMenuBuilderTest.kt:8)
	at io.kotlintest.Spec$runTest$callable$1$1.invoke(Spec.kt:124)
	at io.kotlintest.Spec$runTest$callable$1$1.invoke(Spec.kt:15)
	at io.kotlintest.Spec$runTest$initialInterceptor$1$1.invoke(Spec.kt:116)
	at io.kotlintest.Spec$runTest$initialInterceptor$1$1.invoke(Spec.kt:15)
	at io.kotlintest.Spec.interceptTestCase(Spec.kt:78)
	at io.kotlintest.Spec$runTest$initialInterceptor$1.invoke(Spec.kt:116)
	at io.kotlintest.Spec$runTest$initialInterceptor$1.invoke(Spec.kt:15)
	at io.kotlintest.Spec$runTest$callable$1.call(Spec.kt:124)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by autonomousapps** [comment]
Small update:
1. Tests also fail when run on Jenkins (as expected)
2. If I disable each test with `.config(enabled = false)`, test command still hangs at test 896, but eventually test process continues with note that it skipped 5 tests.

**Comment by Naitbit** [comment]
The test did fail, not because actual result was different from expected value, but because passed expected list was not sorted. `containsInOrder(List)` function could probably use the KDoc explaining why only sorted expected values are allowed. 
For start it could include the comment that is above the `vararg` version to make it clearer what this Matcher does.

**Comment by autonomousapps** [comment]
To clarify, there are two issues.
1. The tests used `containsInOrder(List)`, and the tests passed when run from Android Studio, but failed when run from CLI with gradle as part of the entire suite. Not sure why that would happen. I also think that, if that method does sorting, its name is misleading.
2. Adding a single KotlinTest test imposes a one-time cost of ~44s to the test suite. A second KotlinTest test does not add another 44s. Seems like a bug.

**Comment by sksamuel** [comment]
The second point is the project config scan. You can disable this with
-DprojectConfigScan=false

**Comment by sksamuel** [comment]
And you need 2.0.7 for that to work.

**Comment by autonomousapps** [comment]
Thanks for the note about `-DprojectConfigScan=false`. Can you clarify a bit? What is that? Why would I want it `true`?

**Comment by Naitbit** [comment]
@autonomousapps If I understand https://github.com/kotlintest/kotlintest/blob/7607f81d40a2c0eb15638f0b4925488572f1f845/src/main/kotlin/io/kotlintest/Project.kt#L10
correctly, then KotlinTest tries to find config file in a project via reflection. Since the tested project has none it has to look through all classes. And since the project is huge, and scan is single threaded it takes a while. Flag seems to disable automatic config detection.

**Comment by sksamuel** [comment]
@Naitbit is 100% correct.

I am thinking for 2.1 we should disable the scan by default. What do you think @helmbold

**Comment by autonomousapps** [comment]
What is the value of a config file? If you disable the scan by default,
perhaps it would be better to require users to specify the location of this
file, rather than scanning for it?

Also, thanks for explaining, Sebastian.

**Comment by sksamuel** [comment]
The config file allows you specify interceptors and other things that should apply to all tests.

**Comment by autonomousapps** [comment]
Running the test suite like this:
```
./gradlew -DprojectConfigScan=false testRDebugUnitTest
```
I still had the 45s delay.

I also tried:
```
export projectConfigScan=false
export projectConfigScan="false"
```
Still no impact.

I tried
```
System.setProperty("projectConfigScan", "false")
```
No impact.

I tried adding an empty implementation of `ProjectConfig()`
```
object KotlinTestConfig : ProjectConfig()
```
No impact.

Tried a non-empty implementation:
```
object KotlinTestConfig : ProjectConfig() {
    private var started: Long = 0L

    override fun beforeAll() {
        started = System.currentTimeMillis()
        println("Starting KotlinTest")
    }

    override fun afterAll() {
        val duration = System.currentTimeMillis() - started
        println("Finishing KotlinTest: overall time for test run: " + duration)
    }
}
```
No impact. Moreover, I didn't see the interceptors run. What gives?

**Comment by autonomousapps** [comment]
Update. I can confirm it's actually finding and calling `KotlinTestConfig` -- but it still takes FOREVER to find it.

I also tried adding `System.setProperty("projectConfigScan", "false)` to my `build.gradle`, and then had it print the value of that property in an `afterEvaluate {}` block, but the `Project` class always returns null with the call `System.getProperty("projectConfigScan")` (I debugged and saw this). I am not sure how to get KotlinTest to respect that property. Any advice?

**Comment by helmbold** [comment]
Project scanning per reflection seems indeed to be problematic. But disabling it by default would make our project configuration mechanism somewhat counter intuitive. I could imagine that we enforce a certain class name for the config (instead of looking up some class derived from our config type). What do you think?

**Comment by sksamuel** [comment]
Ok, so enforcing a class called `io.kotlintest.config.ProjectConfig`, which we can lookup using `Class.forName`. Would work, and avoid any lookup penalty.

**Comment by autonomousapps** [comment]
I have no problem with such a constraint.

**Comment by helmbold** [comment]
@sksamuel I'm just wondering if the full packages (io.kotlintest.config) of the qualified name could collide with Java 9 modules. As far as I know split packages (one namespace in two modules) is no longer allowed.

**Comment by sksamuel** [comment]
Ok, we just need a package name that's not in use - io.kotlintest.userconfig or something like that.

**Comment by sksamuel** [comment]
io.kotlintest.provided.ProjectConfig

**Comment by helmbold** [comment]
Perfect.

**Comment by helmbold** [comment]
Will be fixed with https://github.com/kotlintest/kotlintest/tree/feature/junit5

**Comment by christopheblin** [comment]
Could you provide a status update for this issue please

It seems that the junit5 branch is stalled since december so how can we avoid loosing so much time before any test is runned ...

**Comment by sksamuel** [comment]
Are you using 2.0.7 and disabling the project scan ?

**Comment by sksamuel** [comment]
Can you give me a bigger stack trace, so I can fix and release 2.0.8

**Comment by christopheblin** [comment]
[kotlintest_projectConfigScan.txt](https://github.com/kotlintest/kotlintest/files/1789273/kotlintest_projectConfigScan.txt)

These are only the first lines, I can send the complete output if you need more

**Comment by sksamuel** [comment]
Ok that looks sufficient thanks. It's breaking on non-class files in the
jars. Should be an easy fix.

**Comment by sksamuel** [comment]
This is fixed in 3.0.0 RC1.
If you have the time, please give it a try. There are some breaking changes that should be aware of by reading the changelog here: https://github.com/kotlintest/kotlintest/blob/master/CHANGELOG.md
